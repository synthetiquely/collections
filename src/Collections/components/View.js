import React, { Component } from 'react';
import glamorous from 'glamorous';

// const Container = glamorous.div(
//   {
//     // display: 'flex',
//     // flexDirection: 'column',
//     // justifyContent: 'center',
//     // alignItems: 'center',
//     ':hover': {
//       boxShadow: '1px 1px 1px #ccc',
//     },
//     cursor: 'pointer',
//   },
//   ({ size }) => {
//     if (size === 'large') {
//       return {
//         gridColumn: 'span 3',
//         gridRow: 'span 1',
//       };
//     } else if (size === 'medium') {
//       return {
//         gridColumn: 'span 2',
//         gridRow: 'span 1',
//       };
//     }
//     return {
//       gridColumn: 'span 1',
//       gridRow: 'span 1',
//     };
//   },
// );

const Image = glamorous.img(
  {
    minWidth: '100%',
    maxHeight: '150px',
    objectFit: 'cover',
  },
  // ({ imageLoaded }) => {
  //   if (!imageLoaded) {
  //     return {
  //       backgroundRepeat: 'no-repeat',
  //       backgroundImage: 'linear-gradient(to right, #ffdb4d, #ffe78c)',
  //       backgroundSize: '100%, 100%',
  //       backgroundPosition: '0 0',
  //     };
  //   }
  //   return {};
  // },
  ({ size }) => {
    if (size === 'large') {
      return {
        gridColumn: 'span 3',
        gridRow: 'span 1',
      };
    } else if (size === 'medium') {
      return {
        gridColumn: 'span 2',
        gridRow: 'span 1',
      };
    }
    return {
      gridColumn: 'span 1',
      gridRow: 'span 1',
    };
  },
);
export default class View extends Component {
  constructor(props) {
    super(props);
    this.state = {
      imageLoaded: false,
    };
  }

  onLoad(status) {
    this.setState({
      imageLoaded: status,
    });
  }

  render() {
    const {
      title, src, size, onClick,
    } = this.props;

    return (
      <Image
        size={size}
        src={src}
        alt={title}
        onClick={onClick}
        onLoad={() => this.onLoad(true)}
        imageLoaded={this.state.imageLoaded}
      />
    );
  }
}
