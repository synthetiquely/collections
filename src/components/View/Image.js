import glamorous from 'glamorous';

const Image = glamorous.img(
  {
    opacity: '0',
    width: '100%',
    verticalAlign: 'bottom',
    willChange: 'opacity',
    transition: 'opacity 2s',
  },
  ({ loaded }) => {
    if (loaded) {
      return {
        opacity: '1',
      };
    }
    return {};
  },
);

export default Image;
