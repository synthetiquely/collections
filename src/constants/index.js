export const KEY_ARROW_LEFT = 37;
export const KEY_ARROW_RIGHT = 39;
export const SIZE_RATIO = 150;
