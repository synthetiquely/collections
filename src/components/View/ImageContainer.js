import glamorous from 'glamorous';
import { SIZE_RATIO } from '../../constants';

const ImageContainer = glamorous.div(
  {
    margin: '2.5px',
    cursor: 'pointer',
    willChange: 'width, flexGrow',
    ':hover': {
      boxShadow: '1px 1px 1px #ccc',
    },
  },
  ({ height, width }) => {
    if (height && width) {
      const size = Math.floor(width * SIZE_RATIO / height);

      return {
        flexGrow: `${size}`,
        width: `${size}px`,
      };
    }
    return {};
  },
);

export default ImageContainer;
