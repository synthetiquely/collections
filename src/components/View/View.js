import React, { Component } from 'react';
import glamorous from 'glamorous';
import { SIZE_RATIO } from '../../constants';

const ImageContainer = glamorous.div(
  {
    margin: '2.5px',
    cursor: 'pointer',
    willChange: 'width, flexGrow',
    ':hover': {
      boxShadow: '1px 1px 1px #ccc',
    },
  },
  ({ height, width }) => {
    if (height && width) {
      const size = Math.floor(width * SIZE_RATIO / height);

      return {
        flexGrow: `${size}`,
        width: `${size}px`,
      };
    }
    return {};
  },
);

const Image = glamorous.img(
  {
    opacity: '0',
    width: '100%',
    verticalAlign: 'bottom',
    willChange: 'opacity',
    transition: 'opacity 2s',
  },
  ({ loaded }) => {
    if (loaded) {
      return {
        opacity: '1',
      };
    }
    return {};
  },
);

export default class View extends Component {
  constructor(props) {
    super(props);
    this.state = { loaded: false };
    this.width = 0;
    this.height = 0;
    this.img = null;
    this.onLoad = this.onLoad.bind(this);
    this.onError = this.onError.bind(this);
    this.setRef = this.setRef.bind(this);
  }

  onLoad() {
    /**
     * Due to the fact that img width and height will be known in advance
     * as soon as we bring API to our app,
     * we could remove this lines of code.
     */
    this.width = this.img.naturalWidth || this.img.width;
    this.height = this.img.naturalHeight || this.img.height;
    this.setState({
      loaded: true,
    });
  }

  // @TODO: handle error case
  onError() {}

  setRef(img) {
    this.img = img;
  }

  render() {
    const { src, title, onClick } = this.props;

    return (
      <ImageContainer onClick={onClick} width={this.width} height={this.height}>
        <Image
          src={src}
          alt={title}
          innerRef={this.setRef}
          loaded={this.state.loaded}
          onLoad={this.onLoad}
          onError={this.onError}
        />
      </ImageContainer>
    );
  }
}
