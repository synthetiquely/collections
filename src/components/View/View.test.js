import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import View from './View';
import ImageContainer from './ImageContainer';
import { SIZE_RATIO } from '../../constants';

describe('View Component', () => {
  const props = {
    image: {
      height: 300,
      width: 250,
      color: 'red',
      src:
        'https://n6-img-fp.akamaized.net/free-vector/abstract-gray-background-with-lines_1108-73.jpg?size=338&ext=jpg',
      description: 'Test Image',
    },
    onClick: jest.fn(),
  };
  it('should render and match its own snapshot', () => {
    const wrapper = mount(<View {...props} />);
    const tree = renderer.create(<View {...props} />).toJSON();

    expect(wrapper).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('should calculate size based on width and height of an image', () => {
    const wrapper = mount(<View {...props} />);
    const size = Math.round(props.image.width / props.image.height * SIZE_RATIO);

    expect(wrapper.state('size')).toEqual(size);
  });

  it('should handle click event', () => {
    const wrapper = mount(<View {...props} />);

    wrapper.find(ImageContainer).simulate('click');

    expect(props.onClick).toHaveBeenCalled();
  });
  it("should pass color to a container if can'not load an image", () => {
    const wrapper = mount(<View {...props} />);
    wrapper.setState({
      errored: true,
    });

    expect(wrapper.find(ImageContainer).props()).toHaveProperty(
      'color',
      props.image.color,
    );
  });
});
