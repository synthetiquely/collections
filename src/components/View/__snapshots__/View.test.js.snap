// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Component should render and match its own snapshot 1`] = `
<div
  className="css-1ykgg8r"
  onClick={[Function]}
  tabIndex={0}
>
  <img
    alt="Test Image"
    className="css-eob8a0"
    onError={[Function]}
    onLoad={[Function]}
    src="https://n6-img-fp.akamaized.net/free-vector/abstract-gray-background-with-lines_1108-73.jpg?size=338&ext=jpg"
  />
</div>
`;
